File: backend/matching-engine/src/main/java/com/commonsengine/matching/service/WorkerLocationService.java
package com.commonsengine.matching.service;

import com.commonsengine.matching.infrastructure.persistence.WorkerLocationEntity;
import com.commonsengine.matching.infrastructure.persistence.WorkerLocationRepository;
import com.commonsengine.platform.exception.NotFoundException;
import com.commonsengine.platform.geo.GeoPoint;
import com.commonsengine.platform.geo.GeoUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.List;

/**
 * 劳动者位置服务（#63）
 *
 * 职责：
 * 1. 更新劳动者实时位置（upsert）
 * 2. 计算边界框（boundingBox）——将圆形搜索区域转换为矩形，用于 DB 索引预筛
 * 3. findNearbyWorkers——边界框预筛 + Haversine 精筛
 *
 * 两阶段查询优化：
 * - 第一阶段：DB 层用 BETWEEN 边界框过滤（走 lat/lng 索引）
 * - 第二阶段：应用层用 Haversine 精确计算球面距离，排除圆外点
 */
@Service
public class WorkerLocationService {

    private static final Logger log = LoggerFactory.getLogger(WorkerLocationService.class);

    /** 一度纬度 ≈ 111 公里 */
    private static final double METERS_PER_DEGREE_LAT = 111_000.0;

    private final WorkerLocationRepository repository;

    public WorkerLocationService(WorkerLocationRepository repository) {
        this.repository = repository;
    }

    /**
     * 更新（或创建）劳动者位置——upsert 语义。
     *
     * @return 更新后的实体
     */
    @Transactional
    public WorkerLocationEntity updateLocation(String workerId, double lat, double lng,
                                                int activeOrderCount) {
        WorkerLocationEntity entity = repository.findByWorkerId(workerId)
                .orElseGet(() -> {
                    WorkerLocationEntity e = new WorkerLocationEntity();
                    e.setWorkerId(workerId);
                    return e;
                });
        entity.setLat(lat);
        entity.setLng(lng);
        entity.setActiveOrderCount(activeOrderCount);
        entity.setUpdatedAt(Instant.now());
        WorkerLocationEntity saved = repository.save(entity);
        log.debug("劳动者位置已更新 worker={} lat={} lng={} active={}",
                workerId, lat, lng, activeOrderCount);
        return saved;
    }

    @Transactional(readOnly = true)
    public GeoPoint getLocation(String workerId) {
        return repository.findByWorkerId(workerId)
                .map(e -> new GeoPoint(e.getLat(), e.getLng()))
                .orElseThrow(() -> new NotFoundException("WorkerLocation", workerId));
    }

    /**
     * 计算以 (centerLat, centerLng) 为圆心、radiusMeters 为半径的边界框。
     *
     * @return [minLat, maxLat, minLng, maxLng]
     */
    public static double[] boundingBox(double centerLat, double centerLng, double radiusMeters) {
        double latDelta = radiusMeters / METERS_PER_DEGREE_LAT;
        // 经度跨度随纬度变化——cos(lat) 越小（高纬度），一度经度的距离越短
        double cosLat = Math.max(Math.cos(Math.toRadians(centerLat)), 0.01);
        double lngDelta = radiusMeters / (METERS_PER_DEGREE_LAT * cosLat);

        return new double[]{
                centerLat - latDelta,
                centerLat + latDelta,
                centerLng - lngDelta,
                centerLng + lngDelta
        };
    }

    /**
     * 查找圆心 (center) 半径 radiusMeters 内的劳动者。
     *
     * 两阶段：边界框 DB 预筛 → Haversine 精筛。
     */
    @Transactional(readOnly = true)
    public List<WorkerLocationEntity> findNearbyWorkers(double centerLat, double centerLng,
                                                         double radiusMeters) {
        double[] bbox = boundingBox(centerLat, centerLng, radiusMeters);
        List<WorkerLocationEntity> candidates = repository.findInBoundingBox(
                bbox[0], bbox[1], bbox[2], bbox[3]
        );

        GeoPoint center = new GeoPoint(centerLat, centerLng);
        return candidates.stream()
                .filter(e -> {
                    GeoPoint loc = new GeoPoint(e.getLat(), e.getLng());
                    return GeoUtils.isWithinRadius(center, loc, radiusMeters);
                })
                .toList();
    }
}

File: backend/matching-engine/src/main/kotlin/com/commonsengine/matching/service/WorkerLocationService.kt
package com.commonsengine.matching.service

import com.commonsengine.matching.infrastructure.persistence.WorkerLocationEntity
import com.commonsengine.matching.infrastructure.persistence.WorkerLocationRepository
import com.commonsengine.matching.infrastructure.persistence.toDomain
import com.commonsengine.platform.domain.Worker
import com.commonsengine.platform.geo.GeoPoint
import com.commonsengine.platform.geo.GeoUtils
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Instant
import kotlin.math.cos
import kotlin.math.PI

/**
 * 劳动者位置服务
 *
 * 职责：
 * 1. 接收劳动者心跳上报，持久化最新位置
 * 2. 空间检索：给定一个坐标点和半径，返回该范围内的活跃劳动者
 *
 * 检索策略：bounding box 预筛选（DB 层）→ Haversine 精确过滤（应用层）
 * 这个两级过滤保证：DB 查询通用（H2/PostgreSQL 兼容），结果精确。
 */
@Service
open class WorkerLocationService(
    private val repository: WorkerLocationRepository,
) {

    /**
     * 上报/更新劳动者位置（心跳）
     */
    @Transactional
    open fun upsertLocation(
        workerId: String,
        name: String,
        lat: Double,
        lng: Double,
        serviceTypes: String = "",
        rating: Double = 5.0,
        activeOrderCount: Int = 0,
    ) {
        val existing = repository.findById(workerId).orElse(null)
        if (existing != null) {
            existing.name = name
            existing.lat = lat
            existing.lng = lng
            existing.serviceTypesCsv = serviceTypes
            existing.rating = rating
            existing.activeOrderCount = activeOrderCount
            existing.lastSeenAt = Instant.now()
            repository.save(existing)
        } else {
            repository.save(
                WorkerLocationEntity(
                    workerId = workerId,
                    name = name,
                    lat = lat,
                    lng = lng,
                    serviceTypesCsv = serviceTypes,
                    rating = rating,
                    activeOrderCount = activeOrderCount,
                    lastSeenAt = Instant.now(),
                ),
            )
        }
    }

    /**
     * 查找坐标点周围 radiusMeters 范围内的活跃劳动者
     *
     * @param center 搜索中心点
     * @param radiusMeters 搜索半径（米）
     * @param maxActiveOrders 最大活跃订单数（排除满单）
     * @return 精确距离内的劳动者列表（已按距离排序）
     */
    @Transactional(readOnly = true)
    open fun findNearbyWorkers(
        center: GeoPoint,
        radiusMeters: Double,
        maxActiveOrders: Int = 3,
    ): List<Worker> {
        // 1. 计算 bounding box（圆形的外接矩形）
        val box = boundingBox(center, radiusMeters)

        // 2. DB 层预筛选
        val candidates = repository.findInBoundingBox(
            minLat = box.minLat,
            maxLat = box.maxLat,
            minLng = box.minLng,
            maxLng = box.maxLng,
            maxActiveOrders = maxActiveOrders,
        )

        // 3. 应用层 Haversine 精确过滤
        return candidates
            .map { it.toDomain() }
            .filter { worker ->
                GeoUtils.isWithinRadius(center, worker.currentLocation, radiusMeters)
            }
            .sortedBy { GeoUtils.distance(center, it.currentLocation) }
    }

    /**
     * 计算给定坐标和半径的 bounding box（外接矩形）
     *
     * 使用简化的度数转换——对中国纬度范围（约 20-50°）足够精确。
     * 1° 纬度 ≈ 111km，1° 经度 ≈ 111km × cos(lat)
     */
    private fun boundingBox(center: GeoPoint, radiusMeters: Double): BoundingBox {
        val latDelta = radiusMeters / 111_000.0
        val lngDelta = radiusMeters / (111_000.0 * cos(center.lat * PI / 180))

        return BoundingBox(
            minLat = (center.lat - latDelta).coerceIn(-90.0, 90.0),
            maxLat = (center.lat + latDelta).coerceIn(-90.0, 90.0),
            minLng = (center.lng - lngDelta).coerceIn(-180.0, 180.0),
            maxLng = (center.lng + lngDelta).coerceIn(-180.0, 180.0),
        )
    }

    private data class BoundingBox(
        val minLat: Double,
        val maxLat: Double,
        val minLng: Double,
        val maxLng: Double,
    )
}
